### ISSUE TITLE
Support ebooks and pdf export

### ISSUE THREAD

**Issue by mdinger** [azerupi/mdBook#88]
Gitbook supports export to [ebooks and pdfs](https://github.com/GitbookIO/gitbook#output-formats) via [calibre](http://calibre-ebook.com/). This might be easy to hook into.

See also https://github.com/rust-lang/rust-by-example/issues/684 for problems this implementation creates for rustbyexample.

**Comment by azerupi** [comment]
I would like to support pdf and ebook format. I think this could already be developed out of tree if you use the Renderer trait from mdBook. 

I am not sure I want to depend on a full blown Gui tool though. There must surely be a better alternative for that.

**Comment by mdinger** [comment]
Not familiar with many tools like this but [pandoc](http://pandoc.org/) also seems like a plausible option.

**Comment by azerupi** [comment]
Yeah pandoc seems a lot better!

**Comment by asolove** [comment]
Did some exploration on this and seems doable. [Here's the default epub version of the Rust book](https://github.com/azerupi/mdBook/files/85834/sample.epub.zip). Note the chapters out of order and links not working.

To get good output, I think we would need to:
- parse the ToC to get the list of md files, in the right order
- concat and transform the markdown files, replacing file links with internal links
- match the themes with epub versions of the styles

I'm interested in working on this but will be a bit slow.

Useful info here: [Pandoc commands and styling options](http://pandoc.org/epub.html)

**Comment by killercup** [comment]
> - parse the ToC to get the list of md files, in the right order
> - concat and transform the markdown files, replacing file links with internal links

@asolove, I have implemented this (among other transformations) in <https://github.com/killercup/trpl-ebook>, feel free to use my code.

**Comment by asolove** [comment]
@killercup great, thanks!

**Comment by azerupi** [comment]
Great! Thanks for doing this :)

>parse the ToC to get the list of md files, in the right order

This is already done in the Rust code, the `MDBook` struct can be iterated on. If you make a new `Renderer` you have access to that.

>concat and transform the markdown files, replacing file links with internal links

Concatenating the markdown files is also not that hard, I do it for the print page.

Replacing the links could be a little trickier, what should internal links look like for pandoc? I know that pulldown-cmark gives you the ability to transform the parsed markdown events before rendering, but it's not well documented. Maybe link replacing is in it's capabilities. 

Static files, like images, will probably also need some special treatment to be included correctly?

-----

>I'm interested in working on this but will be a bit slow.

That is absolutely no problem, there is no rush. I will assign this issue to you so that others can see you are working on it. If you need any help, feel free to ask here :)

I am also planning on doing a big refactor (#90) to clean up and create a better API. For example, I am thinking about adding a way to poll the `MDBook` struct for specific chapters, etc. This would make it a lot more flexible for Renderers and if I end up doing something like #93. If you have suggestions or requests that might be relevant, post them in #90 so that I / we can brainstorm and come up with a good design :)

**Comment by killercup** [comment]
> Replacing the links could be a little trickier, what should internal links look like for pandoc?

FIY, I'm doing [some regex work](https://github.com/killercup/trpl-ebook/blob/3ec694afa953b97607401977c662db27a80af7f5/src/helpers/normalize.rs) to transform links relative to the doc.rust-lang.org domain and make reference link names unique for the combined markdown file.

**Comment by azerupi** [comment]
>FIY, I'm doing some regex work to transform links relative to the doc.rust-lang.org domain

>```rust
let cross_section_link = Regex::new(r"]\((?P<file>[\w-_]+)\.html\)").unwrap();
output = cross_section_link.replace_all(&output, r"](#sec--$file)");

>let cross_section_ref = Regex::new(r"(?m)^\[(?P<id>.+)\]:\s(?P<file>[^:^/]+)\.html$").unwrap();
output = cross_section_ref.replace_all(&output, r"[$id]: #sec--$file");

>let cross_subsection_link = Regex::new(r"]\((?P<file>[\w-_]+)\.html#(?P<subsection>[\w-_]+)\)").unwrap();
output = cross_subsection_link.replace_all(&output, r"](#$subsection)");

>let cross_subsection_ref = Regex::new(r"(?m)^\[(?P<id>.+)\]:\s(?P<file>[^:^/]+)\.html#(?P<subsection>[\w-_]+)$").unwrap();
output = cross_subsection_ref.replace_all(&output, r"[$id]: #$subsection");
```

Thanks! Does pandoc auto-generate the anchors from the markdown files in those formats? like `#sec--$file`? Or is that also handled by your code?

**Comment by killercup** [comment]
@azerupi I'm pretty sure pandoc generates those. I've had problems before because pandoc generates slugs in a different way than rustdoc.

It should be possible to add a specific id to each header, though. The syntax is `# Header Name {#header-name}` IIRC.

You might also want to look at [adjust_header_level.rs](https://github.com/killercup/trpl-ebook/blob/3ec694afa953b97607401977c662db27a80af7f5/src/helpers/adjust_header_level.rs) and [adjust_reference_names.rs](https://github.com/killercup/trpl-ebook/blob/3ec694afa953b97607401977c662db27a80af7f5/src/helpers/adjust_reference_names.rs).

**Comment by azerupi** [comment]
Ok thanks for all the information, this will probably help @asolove a lot! :)

**Comment by cetra3** [comment]
Not sure if this will help you guys, but I've created a simple rust tool which will collate multiple markdown files into one, resolving internal links and turning them into anchors into documents

We can use this in a pipeline on the way to converting to PDF:
```
mdcollate book-example/src/SUMMARY.md | pulldown-cmark > test.html && wkhtmltopdf test.html test.pdf
```

Code can be found here:
https://github.com/cetra3/mdcollate

Happy to accept any PRs

**Comment by azerupi** [comment]
@cetra3 That is really cool! 
The plan is to make a "renderer" that does everything so that it can be used with the `mdbook build` command. So using a command line tool adds some complications. Have you thought about exposing the functionality as a crate? 

I am not sure I would add a dependency just for that functionality, because there is always the possibility that it will not be maintained actively. But it could be considered if it offers enough useful methods that we wouldn't have to reinvent here.

**Comment by mkpankov** [comment]
I'm also sceptical about Calibre. We use it in [Russian translation of TRPL](https://github.com/kgv/rust_book_ru) and we've come along several problems with EPUB (links are to descriptions in Russian, for reference):
* [EPUB isn't displayed correctly and has broken links](https://github.com/kgv/rust_book_ru/issues/232)
* [EPUB has question marks instead of characters](https://github.com/kgv/rust_book_ru/issues/232#issuecomment-158546374)
* [EPUB has duplicated code blocks](https://github.com/kgv/rust_book_ru/issues/233)
* I also vaguely remember we had to hack styles in order to get better PDF. Not sure if it's necessary or not with Pandoc

**Comment by azerupi** [comment]
Thanks for sharing your experience :)
We will see if pandoc has the same problems, but I think @killercup used it without too much / any problems?

>I also vaguely remember we had to hack styles in order to get better PDF. Not sure if it's necessary or not with Pandoc

I am not sure how this is handled with Pandoc, but having a custom theme could be a good thing.

**Comment by cetra3** [comment]
It's probably possible to wrap up those command line tools into a combined tool or expose it as a rust library.  The last component (html to pdf) would need to use FFI as `wkhtmltopdf` is written in C.  Not sure whether this adds too much dependency on externalities though.

The complication arises in that markdown is a superset of HTML which means that you need something that can present HTML in a printable fashion.  In my experience with this problem, Pandoc and Calibre will do a subset, but you won't get full parity.

**Comment by killercup** [comment]
There are a few things to be aware of, but in general pandoc is really amazing at converting Markdown to LaTeX. Which is what you want, I think—it has some very nice features that you currently can't get with HTML-to-PDF converters. For example, my PDF versions of the Rust Book include cross-references like "This is a mutable variable binding (section 5, page 163)".

If you're no LaTeX wizard (I'm not), you might want to look at [this template I threw together](https://github.com/killercup/trpl-ebook/blob/3ec694afa953b97607401977c662db27a80af7f5/lib/template.tex).

If you have any issues with this, just mention me.

**Comment by azerupi** [comment]
Thanks for all your help Pascal! 
I will definitely look at what you have currently running and I am pretty sure we will end up stealing a lot of your code (if that is ok with you) :wink:

**Comment by gambhiro** [comment]
+1 for the effort, I am looking forward to using `mdbook` to produce ebooks.

It seems to have stalled a bit, is anyone currently working on this?

**Comment by azerupi** [comment]
>It seems to have stalled a bit, is anyone currently working on this?

Indeed, it has stalled a bit. In the last 6 months I have been overwhelmed with work at school :confused: 

I am (very) slowly working on the refactoring / clean-up that I wanted to do. And that work is probably going to change the way this specific feature is going to be implemented. Hopefully I will have some time in September to make significant progress on the internal rewrite so that I can work on new features again.

**Comment by gambhiro** [comment]
@azerupi How much space is there for discussing this feature? There are some specific things I would be looking for in a CLI ebook helper, but maybe you are already determined in which way to go.

Some time ago I wrote [prophecy](https://github.com/profound-labs/prophecy), a ruby gem to automate the tasks I needed when producing ebooks. [This is](https://github.com/profound-labs/forest-path/tree/master/publish/final) and example of the output. It has been very useful for me, but I believe I am the only user :)

I have been wanting to rewrite it with some of the hindsight since its early days, but when I saw this I thought maybe `mdbook` would be able to produce the same results.

There is an [asciinema recording](https://asciinema.org/a/5680) to see to sort of things it does.

**Comment by azerupi** [comment]
I'm open to all ideas :)

**Comment by gambhiro** [comment]
Thanks. I will gather my thoughts and post a longer comment on what ideas worked.

**Comment by gambhiro** [comment]
When I was building prophecy, it was important to be able to:

- A) quickly build a well-designed ebook from minimal input (i.e. just the manuscript in markdown and the TOC sequence)
- B) but let the lib read settings in the book's folder to influence structural behaviour on a per-book basis,
- C) have access to the attributes of the book and the chapters in the manuscript files (such as the ERB-style `<%= chapter.title %>` or inserting the ISBN number with `<%= book.isbn_ebook %>`)
- D) add custom CSS with `@font-face` embedded fonts,
- E) build the EPUB and MOBI with different settings,
- F) generate valid ebooks, trigger no warnings in Sigil's validator
- G) generate both the [toc.ncx][^toc] (for the TOC menu) and an
  [HTML Contents][^htmltoc] page which the reader sees after the title page

[^toc]: https://github.com/profound-labs/prophecy/blob/master/lib/prophecy/assets/epub_template/OEBPS/toc.ncx.erb

[^htmltoc]: https://github.com/profound-labs/prophecy/tree/master/lib/prophecy/generators/book/manuscript/xhtml

I thought it was going to be an insane lot of config options, but after a while
it started to be sufficient for any book, and [this much was enough][^enough].

[^enough]: https://github.com/profound-labs/prophecy/blob/master/lib/prophecy/book.rb

I should add that LaTeX is mentioned a lot in the sources, but I ended up
avoiding to generate LaTeX content. Now this might be a different experience if
you are not so picky about every paragraph.

I produce books for printing, and the LaTeX files have the most specific
hair-trigger accurate tweaks and custom macros, so usually I produce the LaTeX
sources first, and convert to markdown from there for the ebooks.

The book's config files were in `.yml` format, and their option keys would get
overwritten from the general towards the specific.

- `book.yml` - general info (title, author, ISBN)
- `epub_mobi.yml` - shared data for EPUB and MOBI ([chapter list for TOC][^chapters], language attr, etc)
- `epub.yml` - only for epub (such as excluding assets which only go to the mobi)
- `mobi.yml` - only for mobi

[^chapters]: https://github.com/profound-labs/travessia-na-floresta/blob/master/epub_mobi.yml

For `F)`, I wanted to be able to track content changes with a diff tool, but
Sigil had a habit of renaming the folders. Check out the [epub template][^epub]
that worked eventually. The lib also copies `Fonts`, `Images` and `Styles` from
[assets][^assets] into the `OEBPS` folder. See the book [Travessia][^trav] for example.

(btw I lost the habit of committing the generated epub source files. It is
useful to be able to inspect them when working on a book but committing it was
overkill.)

[^trav]: https://github.com/profound-labs/travessia-na-floresta/tree/master/build/epub/OEBPS

[^assets]: https://github.com/profound-labs/prophecy/tree/master/lib/prophecy/assets

[^epub]: https://github.com/profound-labs/prophecy/tree/master/lib/prophecy/assets/epub_template

Although the EPUB format is liberal about folder names for images, fonts, etc.,
I found that when other people made small corrections in the ebook with
[Sigil][^sigil], it would rename the folders without asking. The above template
avoids this.

[^sigil]: https://github.com/Sigil-Ebook/Sigil

For the design in `A)`, [three stylesheets][^css] were necessary, because you

1. want to make it pretty in the EPUB for iPads and such, but
2. for the Kindle screen (Paperwhite and newer) you need to tone it down or optimize for contrast
and readability, and
3. you want to support old Kindles which don't comprehend `@font-face` CSS.

[^css]: https://github.com/profound-labs/prophecy/tree/master/lib/prophecy/assets/Styles

You can see in [page.xhtml.erb][^page] that it would select different stylesheets
depending on whether it was building for EPUB or MOBI.

The new Kindles use the `KF8` format, and you can support the legacy format with
`media="amzn-mobi"` media query, while the new Kindles will take the
`media="amzn-kf8"`.

[^page]: https://github.com/profound-labs/prophecy/blob/master/lib/prophecy/assets/layouts/page.xhtml.erb

To generate MOBI, the best practice seems to be to build it as an EPUB, then run
Amazon's [Kindlegen][^kindlegen] CLI tool on it, to produce the MOBI.

Now Kindlegen has the strange logic of including the source EPUB in the
resulting mobi file, and so your output will be double in size.

The response to this was the [kindlestrip.py][^kindlestrip] script by Paul Durrant, which
strips this out from the mobi, see more on what it does in the comments in its
header.

[^kindlegen]: https://www.amazon.com/gp/feature.html?ie=UTF8&docId=1000765211

[^kindlestrip]: https://github.com/profound-labs/prophecy/blob/master/bin/kindlestrip.py

Back to prophecy, the gem had the stylesheets in the user's gemdir, so the behaviour was:

- initialize a new book without the CSS
- copy from the gemdir when building the ebook files
- allow customizing with a command `prophecy assets`, which made a copy of the assets to the book's folder
- if local assets were found, the lib would compile and use those instead of those in the gemdir.

Some of this wasn't so good. It turns out that almost all ebooks needed at least
a little tweak in either the CSS, the typefaces, etc., and I got into the habit
of always including the assets with the book.

This is probably enough for now. Phew, I hope all the links work! :)

**Comment by gambhiro** [comment]
The details looks a bit scary, but much of the complexity is dealt with in the page templates.

Otherwise it is just this much:

- read in settings
- use local assets when provided
- copy assets to the right folder
- render the ebook data files
- render chapters to HTML
- zip to EPUB
- plus kindlegen, kindlestrip when doing MOBI

**Comment by gambhiro** [comment]
I would like to contribute code too, but I would need guidance. It has not been long since I read my first Rust tutorial :) I would much rather add to a well-designed outline where I can.

**Comment by azerupi** [comment]
@gambhiro Thanks for all info, I glanced over it and will certainly take a deeper look when the time comes to write the pdf / epub renderer!

In the mean time, I updated the [issue for the refactoring](https://github.com/azerupi/mdBook/issues/90) I want to do. It outlines at a high level what I think should happen. This should unblock a lot of the most wanted features. 

I put the points in the order I think they should be implemented, of course this order doesn't need to be followed strictly. 

At this point I think it's important to discuss and iterate over some designs. So if you want to help out I suggest you take a look and start a discussion in the area that interests you :smiley: 

There are a couple of issues that can be implemented directly too, like: 
- [Switch to the log crate](mdBook/issues/151)
- [Switch to Serde](mdBook/issues/18)
- Replace JS and CSS dependencies with their equivalent from npm, to allow more regular updates

I will happily mentor and answer any questions you have, don't be afraid to ask.

**Comment by gambhiro** [comment]
Thanks. I'll level up :)

**Comment by mdinger** [comment]
[Crowbook](https://crates.io/crates/crowbook) just released an [update](http://lise-henry.github.io/articles/0.11.0.html) and I wasn't aware of it at all as a publisher to epub/pdf. Might be worth considering as a renderer for these formats.

**Comment by gambhiro** [comment]
Yes, she is doing very diligent work on that project. I have been working on the ebooks for the past few days, and first thing I did was to read what she wrote in crowbook for this.

It parses the book's files to a data structure, selects the output format from defaults or what the user said on the cli, and gives the data to a function that will do whatever it needs to produce the output files.

Pretty much the obvious thing huh? Basically the task of rendering some markdown files to different formats is a simple and well-defined problem in its nature, so there won't be a lot of different solutions that make sense. It's a lot like the static-site generator idea as well.

One pleasantly surprising thing which @lise-henry does there though, is that she parses the documents into an "Abstract Syntax Language" tree (basically a tree of tokens down to paragraph, link, emphasis) and the output functions iterate over that data. This, instead of the output function reading in the markdown file for its own purposes and doing something with that.

That token tree allows the content to be processed in terms of writing prose or typographical customs, and so she wrote a module that lets crowbook have opinions about the grammar in the text or correct punctuation. She's a novel writer so it makes sense she would be interested in an automated sanity check which the machine can do.

I remember thinking, "hm, why don't I just use crowbook?" and the only convincing point I could come up with is that I like writing the code :) Which is pretty much why she started writing crowbook as well. I think we like to be at that place, writing code to solve a problem we understand.

There is [this TED talk][^talk] by Uber's founder Travis Kalanick, where he is asked why is he doing this. He says, "... the way I like to describe it is it's kind of like a math professor. You know? If a math professor doesn't have hard problems to solve, that's a really sad math professor."

Have you tried to write a basic but efficient [primality test](https://projecteuler.net/problem=10)? I think you would find it satisfying!

Best wishes for the new year!

[^talk]: http://www.ted.com/talks/travis_kalanick_uber_s_plan_to_get_more_people_into_fewer_cars

**Comment by mdinger** [comment]
Regardless, even if you make a library to do such a thing, the task is somewhat orthogonal to the purpose of mdbook. As such, it probably make sense to release it as a crate.

Whether you think it's worth creating something so similar from scratch is up to you. Alternatively, her project might benefit from contributors and additional features.

**Comment by steveklabnik** [comment]
an update: https://github.com/lise-henry/epub-builder

**Comment by Luthaf** [comment]
Would you be interested in a PDF renderer using LaTeX, once #200 is merged? I could try to contribute this. I am using mdbook to build the user manual for my project, and I'd like to have a PDF version of it.

**Comment by gambhiro** [comment]
@Luthaf That would be mighty good. I am also interested in using mdbook for producing LaTeX files. Would you mind opening a separate issue for it and outline what you have in mind?

The options I see is either to depend on the pandoc crate (and hence on the user installing pandoc), or to take a token list as it comes from pulldown-cmark and write an exporter to LaTeX (crowbook does that in [latex.rs](https://github.com/lise-henry/crowbook/blob/master/src/lib/latex.rs)).

Pandoc is very reliable and comprehensive. The user would have to install LaTeX anyway, and at that point it's not a big deal to also install pandoc.

The advantage of writing a token exporter is that it allows for custom typographic processing (recognising the first word of a paragraph, etc.), and I imagine we are not so concerned with that.

**Comment by azerupi** [comment]
@Luthaf Yes, very interested!
I think the first step is to implement a renderer for pulldown-cmark to convert from markdown to latex instead of html. I am not sure how stable the pulldown-cmark API is. This can be done independently of #200 and it would be useful for more than just mdBook :)

**Comment by Luthaf** [comment]
> I think the first step is to implement a renderer for pulldown-cmark to convert from markdown to latex instead of html.

That was basically my idea. pandoc looks nice (I've never used it), but why add another dependency if we can easily go without it? I'll try to get a poc working, and will open an issue here about this.

**Comment by gambhiro** [comment]
>  if we can easily go without it

Tables with headings and rules, smart quoting, tex ligatures, ampersands and other specials, list within quotes... there are so many things to think about.

**Comment by Luthaf** [comment]
> Tables with headings and rules, smart quoting, tex ligatures, ampersands and other specials, list within quotes... there are so many things to think about.

Yep, that true =/. I'll give it a try to grasp the internal of the code, and I'll check how hard this is. If I can not make it work, I'll check how I can make this work on Windows/Linux/OS X with calling user-provided pandoc.

Do you know any test suite for markdown files? Maybe the pandoc's one would be a good listing of all the weird cases.

**Comment by gambhiro** [comment]
I'd say just adopt pandoc. Even if you did without it, the PDF feature is still not stand-alone, because there is always the assumption that the user has gone through the trouble of installing a LaTeX distribution.

There exists a [spec for CommonMark](https://github.com/jgm/CommonMark/) and there is Multimarkdown's [test suite](https://github.com/fletcher/MMD-Test-Suite). 'Simple' is not the right word for it...

**Comment by gambhiro** [comment]
@Luthaf for getting familiar with the mdbook code, I recommend starting with the [multilang](https://github.com/gambhiro/mdBook/tree/multilang) branch. It changes the representation of the book's data than the current master. It will already build books, and if you discover bugs that would help for #200.

**Comment by vandenoever** [comment]
A very naive conversion that works pretty nice for the rust book second edition:
```bash
cd second-edition/src
pandoc -o rust.epub [Sc]*.md a*.md
```

**Comment by dkotrada** [comment]
my 2 cents https://github.com/tectonic-typesetting/tectonic

**Comment by budziq** [comment]
@vandenoever thanks, this seams similar to what @killercup had provided in https://github.com/killercup/trpl-ebook

@dkotrada tectonic looks interesting! I understand that it does require 3'rd party tools to convert TeX/LaTeX to pdf. Am I right?

**Comment by dkotrada** [comment]
@budziq yes you are right.

**Comment by Michael-F-Bryan** [comment]
As a bit of an update on this front, I've started working on adding the ability to use alternative backends (#507). It's still very much in the experimental phase and nothing has been merged yet, but I've already got an [`EPUB` renderer][1] and it generates fairly reasonable outputs. 

Under the hood we use @lise-henry's [epub-builder] crate (awesome crate by the way) to generate an `EPUB` document, with `mdbook` calling the `mdbook-epub` renderer and passing in a copy of the entire book's contents, as well as configuration, via `stdin`. 

This means backends aren't limited to just Rust, for example you could use a python script that passes all Rust snippets to `rustc` or `rustdoc` and makes sure they work (like the builtin `mdbook test` command), with the "rendered output" just being a pass or fail.

[1]: https://github.com/Michael-F-Bryan/mdbook-epub
[epub-builder]: https://github.com/lise-henry/epub-builder

**Comment by dustinmatlock** [comment]
Probably the easiest way: **print** at the top right and **save as PDF**.

![Print Rust PDF](https://i.imgur.com/B1JxihM.png)

**Comment by d8aninja** [comment]
One of the things that doesn't seem to be mentioned anywhere on this ticket is the ability to _highlight the important bits_.  I have used a chrome extension called Hypothesis to do this until recently but a) chrome extension, ew b) its pretty sloppy about whether the highlights are saved under your personal view or public view (ie, in some cases you can see others' highlights) and c) I say recently because I'm pretty sure when the book gets updated, all my highlights and attached notes do, too.  

Anyway, just wanted to add my two cents and support for a PDF version released parallel to the site update.  I realize that must be a lot harder than many of us make it out to be, and you all are doing a wonderful job, regardless of the format in which we are all consuming your work.  Thanks!

**Comment by XVilka** [comment]
Seems that [mdproof](https://github.com/Geemili/mdproof) can be used for such a task

**Comment by mkurnikov** [comment]
Trying to open `print.html` page from the browser gives me out-of-memory, but this one works
```
google-chrome-stable --headless --print-to-pdf=rust_book.pdf https://doc.rust-lang.org/beta/book/print.html
```

**Comment by Binlogo** [comment]
Nice feature request.
Print the pdf with browser is a workaround way, the format looks nice already. 👍
Of course, if mdbook support export feature, it will be convenient for Continuous Integration.

**Comment by heyakyra** [comment]
> if mdbook support export feature, it will be convenient for Continuous Integration.

Yes, I think this can be an important bit. Is this work useful? https://github.com/void-linux/void-docs/pull/416

**Comment by Huy-Ngo** [comment]
I can't install `mdbook-latex` because it depends on `harfbuzz_rs`, which currently can't be compiled for a minor bug (https://github.com/manuel-rhdt/harfbuzz_rs/pull/30). It's fixed but not released.

I think the maintainers would rather encourage user to use external plugins like `mdbook-latex` or `mdbook-epub` rather than internal implementation.

**Comment by ildar** [comment]
As an example on how I'd like a ebook to look like.
Rust by example book: https://flibusta.is/b/619885/

**Comment by Huy-Ngo** [comment]
> As an example on how I'd like a ebook to look like.
> Rust by example book: https://flibusta.is/b/619885/

I see an almost blank page in Russian. Which part of it do you mean you want an ebook to look like?

**Comment by XVilka** [comment]
Since there is zero interest to support that in mdBook, I recommend a relatively new framework to create books, more flexible that commonly known [Bookdown](https://bookdown.org) - [Quarto](https://quarto.org/). It's [pandoc](https://pandoc.org/)-based, thus can export to basically anything. You can see [their gallery](https://quarto.org/docs/gallery/) for samples how such different formats and exports look like. It's quite [actively developed](https://github.com/quarto-dev/quarto-cli) as well.

**Comment by aplatypus** [comment]
@dustinmatlock ... I am afraid that I disagree:

> Probably the easiest way: **print** at the top right and **save as PDF**. Of course, this doesn't solve the EPUB issue, but programming books are sometimes best viewed in PDF. _The book's website looks great on mobile._
> 
> ![Print Rust PDF](https://camo.githubusercontent.com/9e45d3101d29d34d6ac6a26cd972dc3c81a0cddde65b7677ec0298c76ec10666/68747470733a2f2f692e696d6775722e636f6d2f42314a7869684d2e706e67)

A PDF file is usually complete in that a link from the Table of Contents, when clicked would go to page #202 say.  With a generated PDF file from a browser, clicking a hyperlink in the resulting PDF provieds this informative explaination ...
> 
![image](https://user-images.githubusercontent.com/67028/140647499-fc9d464e-ca48-4b1e-8e86-6d5b0b1e6ca3.png)

```
Firefox can’t establish a connection to the server at 127.0.0.1:3000.

    The site could be temporarily unavailable or too busy. Try again in a few moments. ...
```

An export function would at least support hyperlinks for Table of Contents, an Index and Footnotes.  Sometimes it is useful for an exported file to link enternally.. In such cases I believe the PDF format specifies "internal" and "external" links.

Yes, a saved PDF is a usable solution to an off-line document, it is not a solution that I can save to a tablet and use when I'm out of touch with the internet, or late at night, etc.

Have a good one ...!

**Comment by HollowMan6** [comment]
Hi all! I just created a mdBook backend named [mdbook-pdf](https://crates.io/crates/mdbook-pdf) for generating PDF based on [headless chrome](https://github.com/atroche/rust-headless-chrome) and [Chrome DevTools Protocol Page.printToPDF](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF). It depends on Google Chrome / Microsoft Edge / Chromium. The generated page are pretty much alike the one you manually printed to PDF in your browser by opening `print.html` or mentioned here: https://github.com/rust-lang/mdBook/issues/88#issuecomment-541337635 , but with customization of PDF paper orientation, scale of the webpage rendering, paper width and height, page margins, generated PDF page ranges, whether to display header and footer as well as customize their formats, and more, as well as automation. It supports all the platform where Google Chrome / Microsoft Edge / Chromium would work. You can check samples of the generated PDF files in the Artifacts [here](https://github.com/HollowMan6/mdbook-pdf/actions/runs/1766229211).

For the issue aplatypus just mentioned above by using this method https://github.com/rust-lang/mdBook/issues/88#issuecomment-962613257 ,  I guess for those "internal" links inside the book, work should be done in the mdbook side for `print.html` [referring here](https://blog.hubspot.com/marketing/jump-link-same-page) so that all the links linked "internally" would jump inside the generated `print.html`, as all the contents should already be on the `print.html`, there shouldn't be any hyperlinks that jump to other html files in the book. By resolving in this way, the generated PDF would also jump internally instead of opening a browser that won't connect to anything.

**Comment by jacobmellin** [comment]
Hi, I wrote a quick bash script to generate a PDF from mdBook markdown using pandoc and the Eisvogel Pandoc theme. Maybe it is of help to someone:

```bash
#!/bin/sh
# This script converts mdBook markdown output into a pdf using Pandoc/LaTeX and
# the eisvogel pandoc template (https://github.com/Wandmalfarbe/pandoc-latex-template).
# By default, it assumes that the script is put in to a direct subfolder of your
# mdBook project, next to the eisvogel.latex file and your mdBook project root
# contains the book.toml, your markdown sources at ./src and the preprocessed markdown
# will be created in ./book/markdown. Your book.toml file needs to contain the line 
# [output.markdown]
# The path of the resulting pdf file will be ./book/pdf/output.pdf

# Directory that this script is in (e. g. subfolder of PROJECT_DIR)
SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )

# Project directory (contains book.toml, src folder and book output folder)
# Change this if your script is not inside a subfolder (e.g. 'scripts') of the project directory.
PROJECT_DIR="$( dirname $SCRIPT_DIR)"

# Pandoc LaTeX template
# This script works with the eisvogel-Template
# (https://github.com/Wandmalfarbe/pandoc-latex-template).
# If you want to use this template, please put the file
# eisvogel.latex in the same directory as this script
# (e.g. $PROJECT_DIR/scripts/eisvogel.latex).
TPL="$SCRIPT_DIR/eisvogel.latex"

# Build markdown
# Ensure that your book.toml contains the line
# [output.markdown]
mdbook build

# Make output and temp folders
mkdir -p $PROJECT_DIR/book/pdf
mkdir -p $PROJECT_DIR/book/markdown-temp/images

# Copy all images to a single directory
find $PROJECT_DIR/src -name \*.png -exec cp {} $PROJECT_DIR/book/markdown-temp/images \;

# Define output file path
OUTPUT_FILE=$PROJECT_DIR/book/markdown-temp/output.md

# Read meta information from book.toml
CONFIG_FILE_CONTENTS=$( < $PROJECT_DIR/book.toml )

[[ $CONFIG_FILE_CONTENTS =~ title\ +=\ +\"(.*)\" ]] \
    && DOCUMENT_TITLE=${BASH_REMATCH[1]}

[[ $CONFIG_FILE_CONTENTS =~ language\ =\ \"([a-z]*)\" ]] \
    && DOCUMENT_LANGUAGE=${BASH_REMATCH[1]}

[[ $CONFIG_FILE_CONTENTS =~ authors\ +=\ +(\[[^\]]+])\ * ]] \
    && DOCUMENT_AUTHORS=${BASH_REMATCH[1]}

# Write the document title and configuration to output file
cat > $OUTPUT_FILE<< EOF
---
title: ${DOCUMENT_TITLE}
author: ${DOCUMENT_AUTHORS}
date: "11.06.2022"
titlepage: true
fontsize: 10pt
logo: images/laboratorium-logo.png
logo-width: 110mm
toc: true
toc-own-page: true
keywords: [Markdown, Example]
...
EOF

# echo -e "# $DOCUMENT_TITLE\n" >> $OUTPUT_FILE

# Read SUMMARY.md, combine output titles and individual .md file contents
# into single output markdown file
while read line
do
    [[ $line =~ Summary ]] && continue
    # Write SUMMARY.md section titles to markdown file
    # [[ $line =~ ^\# ]] && echo -e "$line\n" >> $OUTPUT_FILE
    
    # Combine different markdown files, increasing the section level
    # for each headline
    # [[ $line =~ \((.*\.md)\) ]] \
        # && sed -e 's/^#/##/g' \
            # "$PROJECT_DIR/book/markdown/${BASH_REMATCH[1]}" \
                # >> $OUTPUT_FILE

    # Combine markdown files, leaving the section headings as they are
    [[ $line =~ \((.*\.md)\) ]] \
        && cat "$PROJECT_DIR/book/markdown/${BASH_REMATCH[1]}" \
            >> $OUTPUT_FILE
        echo -e "\n" >> $OUTPUT_FILE
done < $PROJECT_DIR/src/SUMMARY.md

# Do pandoc conversion of markdown
cd $PROJECT_DIR/book/markdown-temp
pandoc -w latex --template $TPL -o ../pdf/output.pdf output.md --number-sections -V lang=$DOCUMENT_LANGUAGE
```

**Comment by hoijui** [comment]
The main issue when creating a PDF from mdbook sources, is that the Markdown sources are a tree, potentially/likely randomly interlinked (just like HTML, which makes the conversion trivial), while a PDF is a single, linear document.
The main thing to be done there, is to get from a tree to a single, linear document. This can be seen in @jacobmellin s script, for example.
I solved this in MoVeDo (a set of scripts abstracting over multiple tools that take MD sources and produce HTML and/or PDF),
also with a BASH script, considering a few more of the issues (probably not all of them either, though). The script doing this is called [linearize](https://github.com/movedo/MoVeDo/blob/master/scripts/linearize). It uses Pandoc filters to shift header levels, it removes individual files front matters, extracts the titles from there and adds them as headers, prepends header-ids with their sanitized source-file path, and rewrites internal links (links between the MD source files), so they still work within the resulting, single MD file, and maybe one or two additional small things. It uses a file called `doc.yml` as the FrontMatter for the resulting `doc.md`.
The script relies on other scripts and filters inside MoVeDo, but it is by far the most interesting/useful part of the whole piece of software (Maybe the only useful part at all, for anyone but myself). It should probably be extracted/made stand-alone some day.

**Comment by jacobmellin** [comment]
@hoijui Very nice project, I'll definitely check it out.

**Comment by aplatypus** [comment]
@hoijui  ... You could look to the open source Okular tool to see how they load a MD document and render it as PDF.

**Comment by hoijui** [comment]
@aplatypus As I wrote before, the issue is not how to render a single MD file as PDF, that is trivial and possible with *many* tools and libraries. The issue is, how to convert a tree of Markdown files/documents into a single Markdown file.

**Comment by ourongxing** [comment]
Try to use https://github.com/busiyiworld/web-printer/tree/main/packages/mdbook

**Comment by HollowMan6** [comment]
[mdbook-pdf](https://github.com/HollowMan6/mdbook-pdf) now supports Table of Content, see: https://github.com/HollowMan6/mdbook-pdf/issues/1#issuecomment-1366157949

> Hi all! I just created a mdBook backend named [mdbook-pdf](https://crates.io/crates/mdbook-pdf) for generating PDF based on [headless chrome](https://github.com/atroche/rust-headless-chrome) and [Chrome DevTools Protocol Page.printToPDF](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF). It depends on Google Chrome / Microsoft Edge / Chromium. The generated page are pretty much alike the one you manually printed to PDF in your browser by opening `print.html` or mentioned here: [#88 (comment)](https://github.com/rust-lang/mdBook/issues/88#issuecomment-541337635) , but with customization of PDF paper orientation, scale of the webpage rendering, paper width and height, page margins, generated PDF page ranges, whether to display header and footer as well as customize their formats, and more, as well as automation. It supports all the platform where Google Chrome / Microsoft Edge / Chromium would work. You can check samples of the generated PDF files in the Artifacts [here](https://github.com/HollowMan6/mdbook-pdf/actions/runs/1766229211).
> 
> For the issue aplatypus just mentioned above by using this method [#88 (comment)](https://github.com/rust-lang/mdBook/issues/88#issuecomment-962613257) , I guess for those "internal" links inside the book, work should be done in the mdbook side for `print.html` [referring here](https://blog.hubspot.com/marketing/jump-link-same-page) so that all the links linked "internally" would jump inside the generated `print.html`, as all the contents should already be on the `print.html`, there shouldn't be any hyperlinks that jump to other html files in the book. By resolving in this way, the generated PDF would also jump internally instead of opening a browser that won't connect to anything.

**Comment by LegNeato** [comment]
If you are looking for pdf output, check out the project I just posted in https://github.com/rust-lang/mdBook/issues/815#issuecomment-1832689025

**Comment by max-heller** [comment]
I built [`mdbook-pandoc`](https://github.com/max-heller/mdbook-pandoc), a backend powered by [Pandoc](https://pandoc.org). Pandoc is quite mature and supports many output formats, including PDF (I've mainly tested LaTeX) and EPUB. Sample rendered PDF books are [here](https://github.com/max-heller/mdbook-pandoc?tab=readme-ov-file#rendered-books).

**Comment by passion-coder-min** [comment]
I build [mdbookhtml2pdf](https://pypi.org/project/mdbookhtml2pdf/), it can build with mermaid and custom css in html.

**Comment by LunarEclipse363** [comment]
I don't know if this was brought up but it feels like this should be split into separate issues for EPUB and for PDF.

EPUBs are reflowable and effectively a collection of zipped HTML files plus images plus stylesheets. 

PDFs are a fixed-layout format and have very different requirements when it comes to the conversion.

I think the discussion would be much easier to follow if these two formats with essentially opposite goals were not put in one issue.
